Jms connection status

andreas-schroeder opened the pull request:
This is my proposal for JMS connection status.

Fixes #1251

andreas-schroeder commented:
This PR extends #1227, don't merge this before the former is merged/approved.

andreas-schroeder commented:
Commit https://github.com/akka/alpakka/pull/1252/commits/745a6ef7112c2a86abbbe46754ed9899e305a5b4 fixes an issue in the test mentioned here: #1231 - However, a different error message.

andreas-schroeder commented:
rebased on top of master

2m commented:
I am wondering if exposing a boolean as a connectivity status is useful enough.

An alternative would be to expose a `Source[JmsConnectorState, ...]` instead. From that it would be possible for the user to create a stream that consumers state changes and updates an atomic reference if that is what is needed.

WDYT?

andreas-schroeder commented:
The boolean connectivity status would be enough for my use case, as I don't need immediate connectivity updates - having a delay of 1 second would be fine for me.

That said, I see having a stream of connectivity updates could also be beneficial if a fast response to connectivity changes is needed, so let's explore how this could look like.

How would you suggest expose the `Source`? As part of the materialized value, or by using a fan-out shape in the stage?

andreas-schroeder commented:
Hi @2m, I've spiked the change and one challenge is that the source needs to be pre-materialized, and that can only happen during `preStart`. Consequently, the source might not be there at the time the materialized value's `connected` method is called, and I either need to busy wait for the `preStart` to complete or return a `Future[Source[...]]`. I like to keep the API simple, but I guess that to stay in line with the non-blocking mantra of Akka, we need to go for the `Future` - what do you think?

2m commented:
How are you building the status source?

If it is, for example, a [`Source.queue`](https://github.com/akka/akka/blob/c89b4a188412b7184451e01bbf620d868f624bbe/akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala#L623) then you can [prematerialize](https://github.com/akka/akka/blob/c89b4a188412b7184451e01bbf620d868f624bbe/akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala#L92-L99) the source to get out the materialized value, which now can be passed into the graph stage.

Alternatively, to convert from `Future[Source[T, ...]]` to `Source[T, ...]` you can do:

```scala
val fs = Future[Source[T, ...]]
val s: Source[T, ...] = Source.fromFuture(fs).flatMapConcat(identity)
```

andreas-schroeder commented:
@2m here we go - I rebased on master and squashed everything into one commit. Have a look and tell me what you think.

I will add a couple more tests for connection status, and for correct handling of reconnection failures on each stage (creating connection, session, consumer/producer, and destination).

andreas-schroeder commented:
Here are some more tests. I was thinking to replace the status update `Stopping` with `Failed` and `Completed` respectively, but that is a bit more effort than just having `Stopping`. What do you think, @2m ?

andreas-schroeder commented:
So I went ahead and did the split of "stopping" into "failing" and "completing". I've also updated the docs and updated all factory methods in `JmsProducer` and `JmsConsumer`.

For the Java-API, this is a breaking change. 
For the Scala-API, this is a non-breaking change.

This PR is complete from my side, please have a look.

andreas-schroeder commented:
rebased once more on top of master.

andreas-schroeder commented:
Thanks @2m - also for your input, I think this PR wouldn't be in that shape without your ideas.

ennru commented on jms/src/main/scala/akka/stream/alpakka/jms/javadsl/package.scala:
```diff
@@ -0,0 +1,28 @@
+      source: stream.scaladsl.Source[jms.scaladsl.JmsConnectorState, NotUsed]
+  ): Source[JmsConnectorState, NotUsed] =
+    source.map {
+      case Disconnected => JmsConnectorState.Disconnected
```
I guess the Scala states could contain the Java state and offer `asJava`.

andreas-schroeder replied:
Ok, that introduce a dependency from the scaladsl to the javadsl, would that be fine for you?

ennru replied:
Yes, even if the dependency goes the other direction for most cases.

andreas-schroeder replied:
Implemented with https://github.com/akka/alpakka/pull/1252/commits/6a406b743c6209de04bd21e14985a1ea98e512e5

ennru commented:
Thank you for your continued efforts with the JMS connector.

andreas-schroeder commented:
Thank you for keeping everything nicely together! Really appreciate the efforts you and your team put into Alpakka.

Btw, this PR also fixed the test failures mentioned in issues #1031 and #1231.

sullis commented:
when is the next Milestone release?
